pyxem/signals/diffraction_vectors.py

```diff
@@ -215,50 +309,116 @@ def get_magnitude_histogram(self, bins, *args, **kwargs):
 
         return ghis
 
-    def get_unique_vectors(self,
-                           distance_threshold=0):
-        """Obtain the unique diffraction vectors.
+    def get_unique_vectors(self, distance_threshold=0.01,
+                           method='distance_comparison', min_samples=1,
+                           return_clusters=False):
+        """Obtain the unique diffraction vectors by a distance
+        comparison or by clustering using DBSCAN [1].
 
         Parameters
         ----------
         distance_threshold : float
-            The minimum distance between diffraction vectors for them to be
-            considered unique diffraction vectors.
+            The minimum distance between diffraction vectors for them to
+            be considered unique diffraction vectors.
+        method : string
+            The method to use to determine unique vectors. Valid methods
+            are 'distance_comparison' or 'DBSCAN'. 'distance_comparison'
+            iteratively checks the distance between two vectors to
+            determine if they should belong to the same unique vector,
+            and if so, the unique vector is iteratively updated to the
+            average value of the two. 'DBSCAN' relys on the DBSCAN [1]
+            clustering algorithm, and uses an Eucledian distance metric.
+        min_samples : int, optional
+            The minimum number of not identical vectors within one
+            cluster for it to be considered a core sample, i.e. to not
+            be considered noise. Only used for method='DBSCAN'.
+        return_clusters : bool, optional
+            If True (False is default), the DBSCAN clustering result is
+            returned. Only used for method='DBSCAN'.
+
+        References
+        ----------
+        [1] https://scikit-learn.org/stable/modules/generated/sklearn.
+            cluster.DBSCAN.html
 
         Returns
         -------
-        unique_vectors : DiffractionVectors
-            A DiffractionVectors object containing only the unique diffraction
-            vectors in the original object.
+        unique_peaks : DiffractionVectors
+            The unique diffraction vectors obtained by clustering.
+        clusters : DBSCAN
+            The results from the clustering, given as class DBSCAN.
+            Only returned if method='DBSCAN' and return_labels is True.
         """
-        if self.axes_manager.navigation_dimension == 2:
-            gvlist = np.array([self.data[0, 0][0]])
+        # Flatten the array of peaks to reach dimension (n, 2), where n
+        # is the number of peaks.
+        peaks = self.data
+        peaks_flat = self.data.flat
+        peaks_all = np.concatenate(np.array(list(map(
+            lambda x: peaks_flat[x].ravel(),
+            np.arange(peaks.size))))).reshape(-1, 2)
+        if distance_threshold == 0:
+            unique_peaks = np.unique(peaks_all, axis=0)
```

Comment:
This seems like slightly odd behaviour at a glance. I think at the very least we need it to be explicitly stated in the docstring that "if distance_threshold == 0" this happens.

Reply:
> 
> 
> This seems like slightly odd behaviour at a glance. I think at the very least we need it to be explicitly stated in the docstring that "if distance_threshold == 0" this happens.

Is that not what you would expect for distance_threshold=0?

Reply:
It probably is actually - but now it makes me think that we need to reword the docstring to say something like:

"Returns diffraction vectors considered unique by: strict comparison, distance
comparison with a specified threshold, or by clustering using DBSCAN [1]."

I'd then consider making this an - if method=='strict' - statement. And raise a warning if distance_threshold == 0 this is equivalent to a strict determination of uniqueness and this is what you get.

Reply:
> 
> 
> It probably is actually - but now it makes me think that we need to reword the docstring to say something like:
> 
> "Returns diffraction vectors considered unique by: strict comparison, distance
> comparison with a specified threshold, or by clustering using DBSCAN [1]."
> 
> I'd then consider making this an - if method=='strict' - statement. And raise a warning if distance_threshold == 0 this is equivalent to a strict determination of uniqueness and this is what you get.

All right, I'll try to put this in